### ISSUE TITLE
Clarify use case for multi-step or one-pot reactions

### ISSUE THREAD

**Issue by connorcoley** [Open-Reaction-Database/ord-schema#112]
The database needs to be able to accommodate multi-step and one-pot reactions where intermediates are not isolated or characterized between steps. 

__Thinking out loud about some options:__

1. Create new `message`(s) explicitly for this case

1. Require users to define two different `Reaction` protos and link the two by adding a field to a `CompoundPreparation` for `string synthesis_record_id` that, when `preparation.type == SYNTHESIZED`, indicates that this compound came from another reaction example. But how to know which product from the other reaction example? What if they don't even have a guess as to the intermediate's structure? How to indicate whether it is the crude product mixture or the pure species?

1. Require users to define two different `Reaction` protos and link the two by adding a field to `ReactionInput` for `string crude_record_id` that can be used in place of `components`. This indicates that the reaction input is the crude product mixture from `record_id`. How do we handle defining the amount then? If the entire crude mixture is used in a batch synthesis, this is easy. But if only a sample is, we still need access to the amount field.

1. Require users to define two different `Reaction` protos and link the two by adding a new type to `CompoundIdentifier` (`CRUDE`?) and expecting that `string details` would be set to the `record_id` of the prior step. How do we handle the situation where the entire crude mixture is used and so a user can't easily specify the amount?

__Proposal:__
- Add a `CrudeProduct crude` to the `ReactionInput` message. Because multi-step and one-pot syntheses almost always use the full amount, this reflects a more common use case. 
```
message CrudeProduct {
  // The ID of the reaction that led to this crude product mixture.
  string record_id = 1;  
  // The reaction_time of the outcome this mixture corresponds to in 
  // Reaction record_id (by default, we assume the last defined ReactionOutcome).
  Time reaction_time = 2;  
  // Whether workup steps in Reaction record_id were performed prior to use 
  // in this synthetic step (e.g., filtration).
  bool after_workup = 3; 
}
```
- We no longer require `ReactionInput`s to have `len(components) > 0`, as having a defined `crude` is an acceptable alternative.
- We allow reaction records to be connected to preceding synthetic steps through a `repeated string preceding_record_id` field in `ReactionProvenance`. This would be useful even for multi-step syntheses where intermediates _are_ isolated and characterized; perhaps the `details` of a `CompoundPreparation` type `SYNTHESIZED` could be another `record_id` if the preparation is defined in the database.
-  If we ask users to link two `Reaction`s through the `record_id`, we have to be careful about how we (re)assign the values during validation. When we validate a `Dataset` or set of `Reaction`s, we keep track of any placeholder values of `record_id` the user writes. If we encounter that `record_id` used in a later `ReactionInput`, we replace it with the newly-assigned official ID. 
- We no longer require that a `conversion` or `yield` is defined for a reaction.

**Comment by skearnes** [comment]
I'm very hesitant to start trying to link `Reaction` records to each other; it breaks the encapsulation (self-contained nature) of the record and doesn't seem to match the ML use case. I also don't like the idea of having two classes of `Reaction` with different requirements depending on the number of implicit steps...

Possible alternatives:

- Use `addition_order` plus unstructured text to describe everything that went into the pot.
- Add an `Intermediate` message describing the expected intermediate products. This would allow e.g. predicting either (a) the intermediates from the inputs or (b) the products from the intermediates + inputs.
- Instead of linking `Reaction` messages; add additional flexibility to `Compound` (when SYNTHESIZED) that allows for more ambiguous amounts---there will be separate `Reaction` records that could have some link in the `ReactionProvenance` message, but they would be self-contained in the sense that `Reaction` A describes the formation of the intermediates and `Reaction` B describes the formation of the products starting from the intermediates and neither is required to understand the other. I think this is most similar to suggestion (4) above.

**Comment by connorcoley** [comment]
I agree that losing encapsulation would be bad and would complicate the ML use case, but it would allow for more flexibility for capturing general reaction data. 

The `addition_order` and `addition_time` does already give us the flexibility to describe the inputs for multi-step reactions. The primary shortcoming is aligning those additions with multiple distinct temperature/stirring/vessel/etc. conditions. This raises the question of whether we need to move toward `repeated ReactionSetup setups` & `repeated ReactionConditions conditions` fields where both are timestamped (like we are doing outcomes at the moment). This would be a little clunky to me since >90% of entries would only add one entry to these repeated fields.

For the separate reaction message case, I don't know if it will be possible to understand Reaction B without reaction A. 

Perhaps we can bring in @b-shields for comments?

**Comment by skearnes** [comment]
I would be more in favor of restructuring `Reaction` to be more general to support this use case. I don't think repeated fields are so clunky that using separate `Reaction` messages would be better; I'd also be in favor of a broader restructuring.

What about adding a `Step` message that wraps much that is currently in `Reaction`, like inputs, conditions, etc. This would add one level of nesting, avoid introducing any repeated fields (other than `Step`) and avoid needed multiple `Reaction` messages so everything will be self-contained...

**Comment by skearnes** [comment]
Created a draft PR #120; will this solve most of our problems here?

**Comment by connorcoley** [comment]
I might need to mull this over for a while. I like the idea of a `Reaction` being a single synthetic step. Even though adding `ReactionStep` is just one more level, we do still need to think about how to connect steps, define amounts, intermediate purifications but not characterizations, etc.

**Comment by connorcoley** [comment]
Elaborating on idea where each reaction step is a separate `Reaction`
- Second reaction would need to accommodate a special input type that refers to the crude product from a different reaction step (by ID). 
- Might want to define putative intermediate structure
- Might want to allow this to be before/after workup steps

**Comment by connorcoley** [comment]
Note for myself later: with any of these changes, ReactionOutcomes can no longer require a Product or at least one analysis.

**Comment by skearnes** [comment]
Fixed by #218
